File: rust/examples/linux/gyroscope/l3gd20h/demo/src/main.rs
use linux_embedded_hal::I2cdev;
use periph::connection::Connection;
use periph::chips::gyroscope::{L3gd20hFull, ODR_190_HZ, FS_500_DPS};

fn main() {
    let i2c_bus: u8 = std::env::var("I2C_BUS").ok().and_then(|v| v.parse().ok()).unwrap_or(1);
    let addr: u8 = std::env::var("I2C_ADDR")
        .ok()
        .and_then(|v| u8::from_str_radix(v.trim_start_matches("0x"), 16).ok())
        .unwrap_or(0x6A);

    let dev = I2cdev::new(format!("/dev/i2c-{}", i2c_bus)).expect("open i2c bus");
    let conn = Connection::new(dev);
    let mut gyro = L3gd20hFull::new(conn, addr, false).expect("init");

    // --- Configure for shake detection at 190 Hz, ±500 dps ---
    // 190 Hz ODR provides good temporal resolution for shake detection;
    // ±500 dps full scale gives 17.5 mdps/digit sensitivity, suitable for
    // detecting moderate to strong motion without clipping.
    gyro.configure(ODR_190_HZ, 0, FS_500_DPS).expect("configure");

    println!("L3GD20H shake detector running. Shake the device...");

    loop {
        if gyro.data_ready().expect("data_ready") { // Check data ready, () -> Result<bool>
            let (x, y, z) = gyro.gyro().expect("gyro"); // Read angular rate, () -> Result<(f32, f32, f32)>
            let magnitude = libm::sqrtf(x*x + y*y + z*z);
            if magnitude > 1.0 {
                println!("SHAKE DETECTED: mag={:.3} (x={:.3} y={:.3} z={:.3})", magnitude, x, y, z);
            } else {
                println!("x={:.3} y={:.3} z={:.3} mag={:.3}", x, y, z, magnitude);
            }
        }
    }
}

File: cpp/examples/linux/gyroscope/L3gd20h/demo/main.cpp
#include <I2CConnectionLinux.h>
#include <L3gd20h.h>
#include <iostream>
#include <cmath>
#include <unistd.h>

int main() {
    I2CConnectionLinux conn(1, 0x6A);
    L3gd20hFull gyro(conn);

    // --- Configure for shake detection at 190 Hz, ±500 dps ---
    // 190 Hz ODR provides good temporal resolution for shake detection;
    // ±500 dps full scale gives 17.5 mdps/digit sensitivity, suitable for
    // detecting moderate to strong motion without clipping.
    gyro.configure(L3gd20hFull::ODR_190_HZ, 0, 1);  // Configure, (odr 0-3, bw 0-3, full_scale 0-2) -> None

    std::cout << "L3GD20H shake detector running. Shake the device..." << std::endl;

    while (true) {
        if (gyro.data_ready()) {                       // Check data ready, () -> bool
            float x, y, z;
            gyro.gyro(x, y, z);                        // Read angular rate, () -> (float, float, float) rad/s
            float magnitude = std::sqrt(x*x + y*y + z*z);
            if (magnitude > 1.0) {
                std::cout << "SHAKE DETECTED: mag=" << magnitude
                          << " (x=" << x << " y=" << y << " z=" << z << ")" << std::endl;
            } else {
                std::cout << "x=" << x << " y=" << y << " z=" << z
                          << " mag=" << magnitude << std::endl;
            }
        }
    }
    return 0;
}